File: CheapestFlightsWithinKStops/CheapestFlightsWithinKStops.cpp
#include <bits/stdc++.h>
using namespace std;

/*
LeetCode 787 — Cheapest Flights Within K Stops

Dijkstra + (node, stops) state
--------------------------------
Time Complexity:
    O(E * log(V * K))
Space Complexity:
    O(V * K)
*/

class Solution {
public:
    int findCheapestPrice(int n,
                          vector<vector<int>>& flights,
                          int src,
                          int dst,
                          int k) {

        // Build adjacency list
        vector<vector<pair<int,int>>> adj(n);
        for (auto &f : flights) {
            adj[f[0]].push_back({f[1], f[2]});
        }

        const int INF = 1e9;

        // dist[node][edges_used]
        vector<vector<int>> dist(n, vector<int>(k + 2, INF));
        dist[src][0] = 0;

        // (cost, node, edges_used)
        using T = tuple<int,int,int>;
        priority_queue<T, vector<T>, greater<T>> pq;
        pq.push({0, src, 0});

        while (!pq.empty()) {
            auto [cost, node, edges] = pq.top();
            pq.pop();

            // Cheapest valid arrival
            if (node == dst)
                return cost;

            // At most k stops => k+1 edges
            if (edges == k + 1)
                continue;

            // Ignore outdated state
            if (cost > dist[node][edges])
                continue;

            for (auto &[next, price] : adj[node]) {
                int newCost = cost + price;

                if (newCost < dist[next][edges + 1]) {
                    dist[next][edges + 1] = newCost;
                    pq.push({newCost, next, edges + 1});
                }
            }
        }

        return -1;
    }
};


/* --------------------------------------------------
   Driver code & test harness
-------------------------------------------------- */

void run_test(int n,
              vector<vector<int>> flights,
              int src,
              int dst,
              int k,
              int expected)
{
    Solution sol;
    int result = sol.findCheapestPrice(n, flights, src, dst, k);

    cout << "n = " << n << "\n";
    cout << "flights = [";
    for (auto &f : flights) {
        cout << "[" << f[0] << "," << f[1] << "," << f[2] << "]";
    }
    cout << "]\n";

    cout << "src = " << src
         << ", dst = " << dst
         << ", k = " << k << "\n";

    cout << "Expected: " << expected << "\n";
    cout << "Result:   " << result << "\n";
    cout << (result == expected ? "✓ PASS" : "✗ FAIL") << "\n";
    cout << string(60, '-') << "\n";
}


int main() {

    cout << "=== Testing Cheapest Flights Within K Stops ===\n\n";

    // Test Case 1 — LeetCode example
    run_test(
        4,
        {{0,1,100},{1,2,100},{2,3,100},{0,3,500}},
        0, 3, 1,
        500
    );

    // Test Case 2 — cheaper with one stop
    run_test(
        3,
        {{0,1,100},{1,2,100},{0,2,500}},
        0, 2, 1,
        200
    );

    // Test Case 3 — direct only (k = 0)
    run_test(
        3,
        {{0,1,100},{1,2,100},{0,2,500}},
        0, 2, 0,
        500
    );

    // Test Case 4 — unreachable
    run_test(
        3,
        {{0,1,100}},
        0, 2, 2,
        -1
    );

    cout << "\n=== Algorithm Analysis ===\n\n";
    cout << "Time Complexity: O(E * log(V * K))\n";
    cout << "Space Complexity: O(V * K)\n\n";

    cout << "Key Insights:\n";
    cout << " - Stops must be part of the state.\n";
    cout << " - Priority queue guarantees cheapest expansion first.\n";
    cout << " - Early return works because heap is cost-ordered.\n";

    return 0;
}

File: CheapestFlightsWithinKStops/CheapestFlightsWithinKStops.py
from typing import List
from collections import deque

"""
This file contains TWO solutions for:

LeetCode 787 — Cheapest Flights Within K Stops (Medium)

Link:
https://leetcode.com/problems/cheapest-flights-within-k-stops/

Problem summary:
- There are n cities numbered 0..n-1.
- flights[i] = [u, v, price] represents a directed flight.
- Find the cheapest price from src to dst with at most k stops.
- If no such route exists, return -1.

------------------------------------------------------------
Algorithms included:
1) Bellman–Ford (official / safest approach)
2) BFS-level traversal with pruning (common alternative)

Core idea (Bellman–Ford):
- Relax all edges exactly (k+1) times.
- Each iteration allows paths using one more edge.
- Use a temporary array so updates do not leak within the same round.

Core idea (BFS version):
- Process graph level-by-level (each level = one more stop).
- Store (node, cost) in queue.
- Prune using min cost seen so far.
- Often fast, but Bellman–Ford is easier to prove correct.
"""

# ------------------------------------------------------------
# Solution 1 — Bellman–Ford (Recommended / Optimal)
# ------------------------------------------------------------

class SolutionBellmanFord:
    def findCheapestPrice(
        self,
        n: int,
        flights: List[List[int]],
        src: int,
        dst: int,
        k: int
    ) -> int:

        INF = float("inf")
        dist = [INF] * n
        dist[src] = 0

        # k stops => at most k+1 edges
        for _ in range(k + 1):
            ndist = dist[:]  # freeze previous layer

            for u, v, w in flights:
                if dist[u] != INF and dist[u] + w < ndist[v]:
                    ndist[v] = dist[u] + w

            dist = ndist

        return -1 if dist[dst] == INF else dist[dst]


# ------------------------------------------------------------
# Solution 2 — BFS Level Traversal + Pruning
# ------------------------------------------------------------

class SolutionBFS:
    def findCheapestPrice(
        self,
        n: int,
        flights: List[List[int]],
        src: int,
        dst: int,
        k: int
    ) -> int:

        adj = [[] for _ in range(n)]
        for u, v, c in flights:
            adj[u].append((v, c))

        minCost = [float("inf")] * n
        minCost[src] = 0

        q = deque([(src, 0)])
        stops = 0

        while q and stops <= k:
            level_size = len(q)

            for _ in range(level_size):
                node, cost = q.popleft()

                for nxt, price in adj[node]:
                    newCost = cost + price

                    if newCost >= minCost[nxt]:
                        continue

                    minCost[nxt] = newCost
                    q.append((nxt, newCost))

            stops += 1

        return -1 if minCost[dst] == float("inf") else minCost[dst]


# ------------------------------------------------------------
# Driver code & test harness
# ------------------------------------------------------------

def run_test(n, flights, src, dst, k, expected):
    bellman = SolutionBellmanFord()
    bfs = SolutionBFS()

    result_bellman = bellman.findCheapestPrice(n, flights, src, dst, k)
    result_bfs = bfs.findCheapestPrice(n, flights, src, dst, k)

    print(f"n = {n}")
    print(f"flights = {flights}")
    print(f"src = {src}, dst = {dst}, k = {k}")
    print(f"Expected:           {expected}")
    print(f"Bellman-Ford result {result_bellman}")
    print(f"BFS result          {result_bfs}")

    ok1 = result_bellman == expected
    ok2 = result_bfs == expected

    print("Bellman-Ford:", "✓ PASS" if ok1 else "✗ FAIL")
    print("BFS:", "✓ PASS" if ok2 else "✗ FAIL")
    print("-" * 60)


if __name__ == "__main__":
    print("=== Testing Cheapest Flights Within K Stops ===\n")

    # Test Case 1 — LeetCode example
    run_test(
        n=4,
        flights=[[0,1,100],[1,2,100],[2,3,100],[0,3,500]],
        src=0,
        dst=3,
        k=1,
        expected=500
    )

    # Test Case 2 — cheaper with one stop
    run_test(
        n=3,
        flights=[[0,1,100],[1,2,100],[0,2,500]],
        src=0,
        dst=2,
        k=1,
        expected=200
    )

    # Test Case 3 — direct only (k=0)
    run_test(
        n=3,
        flights=[[0,1,100],[1,2,100],[0,2,500]],
        src=0,
        dst=2,
        k=0,
        expected=500
    )

    # Test Case 4 — unreachable
    run_test(
        n=3,
        flights=[[0,1,100]],
        src=0,
        dst=2,
        k=2,
        expected=-1
    )

    print("\n=== Algorithm Analysis ===\n")

    print("Bellman–Ford:")
    print("  Time Complexity: O(E * (k+1))")
    print("  Space Complexity: O(n)")
    print("  Key Insight: Each round allows one more edge,")
    print("               making correctness easy to prove.")

    print("\nBFS + Pruning:")
    print("  Average Performance: Often fast in practice")
    print("  Space Complexity: O(n + E)")
    print("  Caveat: Uses cost pruning; Bellman–Ford is safer")
    print("          for strict correctness guarantees.")
